Add Docs content (#58)
CrowdStrike/ember-headless-form: docs/validation/timing.md

---
title: Timing
order: 4
---

# Timing of dynamic validation

> In the most basic way validation will happen whenever you try to submit a form. This is the way how server-rendered apps without any native validation or JavaScript worked for a long time. And in fact, ember-headless-form enforces validation whenever you try to submit the form.

Reviewer: ```suggestion
In the most basic way validation will happen whenever you try to submit a form. This is how server-rendered apps without any native validation or JavaScript worked for a long time. And in fact, ember-headless-form enforces validation whenever you try to submit the form.
```

But in addition to that, as our app is running client-side, we can do validation even earlier, for example whenever an input was changed. This is what we call dynamic validation here.

> By default validation will happen:

Reviewer 2: Two suggestions:
1. move this section (line 12) up to below the header (line 6)
2. formatting:

```suggestion
By default validation will happen:
- **initially**, when you try to submit
- on **"revalidation"**, when a field is in an invalid state (due to a prior submission attempt), and the user changes its value (a `change` event is fired from the control)
```

In the following example, when you enter an invalid email address like `foo`, nothing will happen until you press the submit button. When pressing the submit button the email field will become invalid and show the (native) error message. When you fix it to be a real email _and_ move focus out of the input (which triggers a `change` event), you will see that the validation error is immediately removed:

```hbs preview-template
<HeadlessForm as |form|>
  <form.Field @name='email' as |field|>
    <div class='my-2 flex flex-col'>
      <field.Label>Email</field.Label>
      <field.Input @type='email' required class='border rounded px-2' />
      <field.Errors />
    </div>
  </form.Field>

  <button type='submit'>Submit</button>
</HeadlessForm>
```

## `@validateOn`

You can customize the first part of _initial_ validation by passing the `@validateOn` argument to the form. It accepts one of these values: (sorted in order how frequently the events happen)

- `submit`: validate when submitting the form. This is the default.

> - `focusout`: validate whenever a control looses focus.

Reviewer: ```suggestion
- `focusout`: validate whenever a control loses focus.
```

- `change`: validate whenever a control triggers a `change` event.

>   Important to note: when a control emits a `change` event depends on its type: for selection-based controls like checkboxes, radios or selects, that happens whenever the selected value changes. For text-based inputs this only happens after loosing focus! See [change event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)

Reviewer: ```suggestion
  Important to note: when a control emits a `change` event, it depends on its type: for selection-based controls like checkboxes, radios or selects, that happens whenever the selected value changes. For text-based inputs this only happens after losing focus! See [change event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)
```
